HeavyFlavorAnalysis/SpecificDecay/plugins/BPHWriteSpecificDecay.cc

```diff
@@ -0,0 +1,683 @@
+#include "HeavyFlavorAnalysis/RecoDecay/interface/BPHRecoCandidate.h"
+#include "HeavyFlavorAnalysis/RecoDecay/interface/BPHPlusMinusCandidate.h"
+#include "HeavyFlavorAnalysis/RecoDecay/interface/BPHMomentumSelect.h"
+#include "HeavyFlavorAnalysis/RecoDecay/interface/BPHVertexSelect.h"
+#include "HeavyFlavorAnalysis/RecoDecay/interface/BPHTrackReference.h"
+
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHMuonPtSelect.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHMuonEtaSelect.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHParticlePtSelect.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHParticleNeutralVeto.h"
+#include "HeavyFlavorAnalysis/RecoDecay/interface/BPHMultiSelect.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHMassSelect.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHChi2Select.h"
+
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHOniaToMuMuBuilder.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHKx0ToKPiBuilder.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHPhiToKKBuilder.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHBuToJPsiKBuilder.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHBsToJPsiPhiBuilder.h"
+#include "HeavyFlavorAnalysis/SpecificDecay/interface/BPHBdToJPsiKxBuilder.h"
+
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+
+#include "DataFormats/PatCandidates/interface/GenericParticle.h"
+#include "DataFormats/PatCandidates/interface/CompositeCandidate.h"
+
+#include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
+#include "TrackingTools/PatternTools/interface/TwoTrackMinimumDistance.h"
+
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include <set>
+#include <string>
+#include <iostream>
+#include <fstream>
+
+using namespace std;
+
+#define SET_LABEL(NAME,PSET) ( NAME = getParameter( PSET, #NAME ) )
+// SET_LABEL(xyz,ps);
+// is equivalent to
+// xyz = getParameter( ps, "xyx" )
+
+BPHWriteSpecificDecay::BPHWriteSpecificDecay( const edm::ParameterSet& ps ) {
+
+  usePV = ( SET_LABEL( pVertexLabel, ps ) != "" );
+  usePM = ( SET_LABEL( patMuonLabel, ps ) != "" );
+  useCC = ( SET_LABEL( ccCandsLabel, ps ) != "" );
+  usePF = ( SET_LABEL( pfCandsLabel, ps ) != "" );
+  usePC = ( SET_LABEL( pcCandsLabel, ps ) != "" );
+  useGP = ( SET_LABEL( gpCandsLabel, ps ) != "" );
+  SET_LABEL( oniaName, ps );
+  SET_LABEL(   sdName, ps );
+  SET_LABEL(   ssName, ps );
+  SET_LABEL(   buName, ps );
+  SET_LABEL(   bdName, ps );
+  SET_LABEL(   bsName, ps );
+  writeMomentum = true;
+  writeVertex   = true;
+
+  if ( ps.exists(             "writeMomentum" ) )
+                               writeMomentum =
+       ps.getParameter<bool>( "writeMomentum" );
+  if ( ps.exists(             "writeVertex"   ) )
```

Comment:
PSet::exists is discouraged.
Use ::fillDescriptions instead
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp